include/swift/AST/Expr.h

```diff
@@ -3557,6 +3558,12 @@ class SerializedAbstractClosureExpr : public SerializedLocalDeclContext {
 /// \endcode
 class ClosureExpr : public AbstractClosureExpr {
 
+  /// The range of the brackets of the capture list, if present.
+  SourceRange BracketRange;
+    
+  /// The (possibly null) VarDecl captured with the literal name "self".
+  VarDecl *CapturedSelfDecl;
```

Comment:
Oh, one more suggestion: Since the question of "why didn't you implement this by reading the capture list?" has come up a few times from different people, it might make sense to include a comment explaining why you're using the `capturedSelfContext` member variable instead. That way people will know the reason just by reading the code.

Reply:
Sounds good!